App crashes when editing custom expression referencing metric when dependent metadata is not loaded

### Describe the bug

Original report: https://metaboat.slack.com/archives/C5XHN8GLW/p1724161690436689
Got a failure in CI: https://github.com/metabase/metabase/actions/runs/10471883147/job/29000500799?pr=47009
The failure is "correct", test is not flaky.

![image](https://github.com/user-attachments/assets/0be412fd-bebf-4c48-858e-f662150ec5bf)

```
app-main.hot.bundle.js:714521 Error: metric with ID: 5 does not exist
    at formatMetric (app-main.hot.bundle.js:235066:15)
    at format (app-main.hot.bundle.js:235019:16)
    at transformPropsToState (app-main.hot.bundle.js:533945:95)
    at new ExpressionEditorTextfield (app-main.hot.bundle.js:534216:22)
    at constructClassInstance (vendor.hot.bundle.js:267178:18)
    at updateClassComponent (vendor.hot.bundle.js:272538:5)
    at beginWork (vendor.hot.bundle.js:274461:16)
    at beginWork$1 (vendor.hot.bundle.js:280276:14)
    at performUnitOfWork (vendor.hot.bundle.js:279407:12)
    at workLoopSync (vendor.hot.bundle.js:279316:5)
```


### To Reproduce

1. New > Metric > Orders > Count > Save
2. New > Question > Orders > Summarize by metric created in previous step > Save > Stay in notebook editor
3. Throttle your network or block `/api/card/:id/query_metadata` URL in network tab in developer tools
4. Refresh the page
5. Before GET `/api/card/:id/query_metadata` request completes, try to edit the aggregation's custom expression

App crashes


### Information about your Metabase installation

master, 63c074d2a226c459141be5ed548ee464a5267806


### Severity

P2

--- a/e2e/test/scenarios/metrics/reproductions/metrics-reproductions.cy.spec.ts
+++ b/e2e/test/scenarios/metrics/reproductions/metrics-reproductions.cy.spec.ts
@@ -1,0 +1,45 @@
+import { SAMPLE_DATABASE } from "e2e/support/cypress_sample_database";
+import { createQuestion, restore } from "e2e/support/helpers";
+
+const { ORDERS_ID, ORDERS } = SAMPLE_DATABASE;
+
+describe("issue 47058", () => {
+  it("should show the loading page while the question metadata is being fetched (metabase#47058)", () => {
+    restore();
+    cy.signInAsNormalUser();
+    cy.intercept("GET", "/api/card/*/query_metadata").as("metadata");
+
+    createQuestion({
+      name: "Metric 47058",
+      type: "metric",
+      query: {
+        "source-table": ORDERS_ID,
+        aggregation: [["count"]],
+      },
+    }).then(({ body: { id: metricId } }) => {
+      createQuestion({
+        name: "Question 47058",
+        type: "question",
+        query: {
+          "source-table": ORDERS_ID,
+          fields: [
+            ["field", ORDERS.ID, {}],
+            ["field", ORDERS.TOTAL, {}],
+          ],
+          aggregation: [["metric", metricId]],
+          limit: 1,
+        },
+      }).then(({ body: { id: questionId, name } }) => {
+        cy.visit(`/question/${questionId}/notebook`);
+
+        cy.findByText("Loading...").should("be.visible");
+        cy.findByText(name).should("not.exist");
+
+        // Only render the notebook editor after the metadata is loaded
+        cy.wait("@metadata");
+        cy.findByText("Loading...").should("not.exist");
+        cy.findByText(name).should("be.visible");
+      });
+    });
+  });
+});

--- a/e2e/test/scenarios/metrics/reproductions/metrics-reproductions.cy.spec.ts
+++ b/e2e/test/scenarios/metrics/reproductions/metrics-reproductions.cy.spec.ts
@@ -1,10 +1,15 @@
 import { SAMPLE_DATABASE } from "e2e/support/cypress_sample_database";
-import { createQuestion, restore } from "e2e/support/helpers";
+import {
+  createQuestion,
+  getNotebookStep,
+  main,
+  restore,
+} from "e2e/support/helpers";
 
 const { ORDERS_ID, ORDERS } = SAMPLE_DATABASE;
 
 describe("issue 47058", () => {
-  it("should show the loading page while the question metadata is being fetched (metabase#47058)", () => {
+  before(() => {
     restore();
     cy.signInAsNormalUser();
     cy.intercept("GET", "/api/card/*/query_metadata", req =>
@@ -18,7 +23,7 @@ describe("issue 47058", () => {
         "source-table": ORDERS_ID,
         aggregation: [["count"]],
       },
-    }).then(({ body: { id: metricId, name: metricName } }) => {
+    }).then(({ body: { id: metricId } }) => {
       createQuestion({
         name: "Question 47058",
         type: "question",
@@ -33,20 +38,26 @@ describe("issue 47058", () => {
         },
       }).then(({ body: { id: questionId } }) => {
         cy.visit(`/question/${questionId}/notebook`);
-
-        cy.findByText("Loading...").should("be.visible");
-        cy.findByText(metricName).should("not.exist");
-        cy.findByText("[Unknown Metric]").should("not.exist");
-
-        cy.wait("@metadata");
-        cy.log(
-          "Only renders the notebook editor (with the summarize button that has the metrics' name on it) after the metadata is loaded",
-        );
-
-        cy.findByText("Loading...").should("not.exist");
-        cy.findByText("[Unknown Metric]").should("not.exist");
-        cy.findByText(metricName).should("be.visible");
       });
+    });
+  });
+
+  it("should show the loading page while the question metadata is being fetched (metabase#47058)", () => {
+    main().within(() => {
+      cy.findByText("Loading...").should("be.visible");
+      getNotebookStep("summarize").should("not.exist");
+
+      cy.findByText("[Unknown Metric]").should("not.exist");
+
+      cy.wait("@metadata");
+      cy.log(
+        "Only renders the notebook editor (with the summarize button that has the metrics' name on it) after the metadata is loaded",
+      );
+
+      cy.findByText("Loading...").should("not.exist");
+      getNotebookStep("summarize").should("be.visible");
+
+      cy.findByText("[Unknown Metric]").should("not.exist");
     });
   });
 });

--- a/e2e/test/scenarios/metrics/reproductions/metrics-reproductions.cy.spec.ts
+++ b/e2e/test/scenarios/metrics/reproductions/metrics-reproductions.cy.spec.ts
@@ -9,7 +9,7 @@ import {
 const { ORDERS_ID, ORDERS } = SAMPLE_DATABASE;
 
 describe("issue 47058", () => {
-  before(() => {
+  beforeEach(() => {
     restore();
     cy.signInAsNormalUser();
     cy.intercept("GET", "/api/card/*/query_metadata", req =>

--- a/e2e/test/scenarios/metrics/reproductions/metrics-reproductions.cy.spec.ts
+++ b/e2e/test/scenarios/metrics/reproductions/metrics-reproductions.cy.spec.ts
@@ -7,7 +7,9 @@ describe("issue 47058", () => {
   it("should show the loading page while the question metadata is being fetched (metabase#47058)", () => {
     restore();
     cy.signInAsNormalUser();
-    cy.intercept("GET", "/api/card/*/query_metadata").as("metadata");
+    cy.intercept("GET", "/api/card/*/query_metadata", req =>
+      req.continue(() => new Promise(resolve => setTimeout(resolve, 1000))),
+    ).as("metadata");
 
     createQuestion({
       name: "Metric 47058",
@@ -16,7 +18,7 @@ describe("issue 47058", () => {
         "source-table": ORDERS_ID,
         aggregation: [["count"]],
       },
-    }).then(({ body: { id: metricId } }) => {
+    }).then(({ body: { id: metricId, name: metricName } }) => {
       createQuestion({
         name: "Question 47058",
         type: "question",
@@ -29,16 +31,21 @@ describe("issue 47058", () => {
           aggregation: [["metric", metricId]],
           limit: 1,
         },
-      }).then(({ body: { id: questionId, name } }) => {
+      }).then(({ body: { id: questionId } }) => {
         cy.visit(`/question/${questionId}/notebook`);
 
         cy.findByText("Loading...").should("be.visible");
-        cy.findByText(name).should("not.exist");
+        cy.findByText(metricName).should("not.exist");
+        cy.findByText("[Unknown Metric]").should("not.exist");
 
-        // Only render the notebook editor after the metadata is loaded
         cy.wait("@metadata");
+        cy.log(
+          "Only renders the notebook editor (with the summarize button that has the metrics' name on it) after the metadata is loaded",
+        );
+
         cy.findByText("Loading...").should("not.exist");
-        cy.findByText(name).should("be.visible");
+        cy.findByText("[Unknown Metric]").should("not.exist");
+        cy.findByText(metricName).should("be.visible");
       });
     });
   });

